astrokat/noisediode.py

```diff
@@ -9,257 +9,419 @@
     from katcorelib import user_logger
 except ImportError:
     from .simulate import user_logger
+from . import _DEFAULT_LEAD_TIME
+from . import max_cycle_len
 
 
-_DEFAULT_LEAD_TIME = 5.0  # lead time [sec]
+def _get_max_cycle_len(kat):
+    if not kat.dry_run:
+        return max_cycle_len(kat.sensor.sub_band.get_value())
+    else:
+        return max_cycle_len('l')
+
+
+def _eval_lead_time_(lead_time):
+    # if not given, apply default value
+    lead_time = float(lead_time or _DEFAULT_LEAD_TIME)
+    user_logger.trace('TRACE: Add lead time {}s to ND'
+                      .format(lead_time))
+    return lead_time
+
+
+def _set_time_(lead_time):
+    # some calls may cause lead_time=None
+    lead_time = _eval_lead_time_(lead_time)
+    timestamp = time.time() + lead_time
+    return timestamp
+
+
+def _set_dig_nd_(kat,
+                 timestamp,
+                 nd_setup=None,  # no pattern set
+                 switch=0,  # all nd off
+                 cycle=False):
+
+    if nd_setup is not None:
+        # selected antennas for nd pattern
+        nd_antennas = nd_setup['antennas'].split(",")
+        # nd pattern length [sec]
+        cycle_length = nd_setup['cycle_len']
+        # on fraction of pattern length [%]
+        on_fraction = nd_setup['on_frac']
+        msg = ('Repeat noise diode pattern every {} sec, '
+               'with {} sec on and apply pattern to {}'
+               .format(cycle_length,
+                       float(cycle_length) * float(on_fraction),
+                       nd_antennas))
+        user_logger.info(msg)
+    else:
+        nd_antennas = [ant.name for ant in kat.ants]
+        cycle_length = 1.
+        on_fraction = switch
+
+    # Noise diodes trigger is evaluated per antenna
+    timestamps = []
+    for ant in sorted(nd_antennas):
+        ped = getattr(kat, ant)
+        try:
+            reply = ped.req.dig_noise_source(timestamp,
+                                             on_fraction,
+                                             cycle_length)
+        except ValueError as e:
```

Comment:
This is something that has been noted on site during observations.
Sometimes the return message from the noise diode does not get through and this caused and error in the previous version which simply had the reply statement.

From the previous times this happened it seems that it is simply the message that got lost and the noise diode did what it was instructed to
But the missing message causes an error that interrupted the observation

Currently, this is not identified as a fatal error (if the message does not come through) -- so the updated implementation does not interrupt the observation, only communicated to the user.

I thought this to be a better implementation then some sleep time guestimate, since it will allow the observation to run, but also inform the user so we can get feedback if this type of errors should be handled differently.

Reply:
OK, although I don't know why it would come out as ValueError.  Good to continue observation.

Reply:
I think I am starting to understand the issue of the `ValueError` -- I went back to find some of the old emails that these were attached to.
Lemme work through them and expand this chat

Reply:
@ajoubertza I found the `ValueError` exception (it was one of the long run galactic observations http://10.97.1.13:8081/tailtask/20191108-0025/progress)

Please advice if I made a mistake in my implementation....

```
2019-11-16 18:23:33.983Z INFO     Pattern set as 0.0 sec ON for 1.0 sec cycle length
Unhandled exception <type 'exceptions.ValueError'> : could not convert string to float: Request dig-noise-source timed out after 30.000644 seconds.
Traceback (most recent call last):
  File "/usr/local/bin/astrokat-observe.py", line 7, in <module>
    sys.exit(main(sys.argv[1:]))
  File "/usr/local/lib/python2.7/dist-packages/astrokat/observe_main.py", line 705, in main
    run_observation(opts, kat)
  File "/usr/local/lib/python2.7/dist-packages/astrokat/observe_main.py", line 265, in __exit__
    noisediode.off(self.array)
  File "/usr/local/lib/python2.7/dist-packages/astrokat/noisediode.py", line 128, in off
    timestamp = _katcp_reply_to_log_(replies)
  File "/usr/local/lib/python2.7/dist-packages/astrokat/noisediode.py", line 21, in _katcp_reply_to_log_
    timestamp = _nd_log_msg_(ant, reply, informs)
  File "/usr/local/lib/python2.7/dist-packages/astrokat/noisediode.py", line 37, in _nd_log_msg_
    actual_time = float(reply.arguments[1])
2019-11-16 18:23:33.991Z kat.katcorelib.utility INFO _install(client.py:693) Client connect loop for ('10.97.1.11', 2060) finished.
2019-11-16 18:23:33.996Z kat.katcorelib.utility INFO _install(client.py:693) Client connect loop for ('10.97.1.11', 2026) finished.
Running cleanup for sched
Running cleanup for katpool
Task 20191108-0025 entered state 'FAILED' with reason: Non-zero script process exit status: 1.
```

Reply:
Thanks for the logs.  So the KATCP request timed out, but the old code still called `_nd_log_msg_`.  The reply arguments would probably some thing like "fail" and "Request timed-out...".  So casting the "timed-out" to a float fails, raising the `ValueError`.

Note that it wasn't `ped.req.dig_noise_source` that raised the exception - it was the `_nd_log_msg_`, called by `_katcp_reply_to_log_`.  We don't need the try..except block around `ped.req.dig_noise_source`.  Also, the improved implementations of `_katcp_reply_` and `_nd_log_msg_` should handle a timed-out request just fine.

Reply:
Thank you for the advise and explanation @ajoubertza !
I have removed the try/except block checking for the ValueError, since the implementation was not well understood